src/hid/hid_device.cc

```diff
@@ -231,45 +226,51 @@ Status HidDevice::ExchangeCbor(Command command,
+Status HidDevice::SendCtapHid(const std::vector<uint8_t>& payload,
+                              std::vector<uint8_t>* response) const {
+  if (payload.empty() || payload.size() - 1 > kMaxDataSize) {
+    return Status::kErrInvalidLength;
+  }
 
+  uint8_t cmd = Frame::kTypeInitMask | payload[0];
+  std::vector<uint8_t> send_data(payload.begin() + 1, payload.end());
   std::vector<uint8_t> recv_data;
-  OK_OR_RETURN(ReceiveCommand(kReceiveTimeout, &cmd, &recv_data));
-
-  // The answer might also be a keepalive.
-  bool has_sent_prompt = false;
-  while (cmd == kCtapHidKeepalive) {
-    KeepaliveStatus keepalive_response = ProcessKeepalive(recv_data);
-    if (keepalive_response == KeepaliveStatus::kStatusError)
-      return Status::kErrOther;
-    if (keepalive_response == KeepaliveStatus::kStatusUpNeeded &&
-        !has_sent_prompt) {
-      has_sent_prompt = true;
-      if (!tracker_->IsTouchPromptIgnored()) {
-        PromptUser();
-      }
-    }
-    OK_OR_RETURN(ReceiveCommand(kReceiveTimeout, &cmd, &recv_data));
+  // Cancel requires no response.
+  if (cmd == kCtapHidCancel) {
+    return SendRecvCommand(cmd, send_data, &recv_data, false, false, true,
+                           false);
   }
-
-  if (cmd != kCtapHidCbor) return Status::kErrInvalidCommand;
-  if (recv_data.empty()) return Status::kErrInvalidLength;
-
-  response_cbor->insert(response_cbor->end(), recv_data.begin() + 1,
-                        recv_data.end());
-
-  if (has_sent_prompt && !expect_up_check) {
-    tracker_->AddObservation("A prompt was sent unexpectedly.");
+  // Wink and Lock require response with empty data at success.
+  else if (cmd == kCtapHidWink || cmd == kCtapHidLock) {
+    return SendRecvCommand(cmd, send_data, &recv_data, true, false, true,
+                           false);
   }
-  if (!has_sent_prompt && expect_up_check) {
-    tracker_->AddObservation(
-        "A prompt was expected, but not performed. Sometimes it is just not "
-        "recognized if performed too fast.");
+  // Init requires non-empty response (cid) and no status code is return.
+  else if (cmd == kCtapHidInit) {
+    return SendRecvCommand(cmd, send_data, &recv_data, true, false, false,
+                           false);
+  }
+  // Ping requires response and no status code is return.
+  else if (cmd == kCtapHidPing) {
+    bool expect_empty_data = send_data.empty();
+    return SendRecvCommand(cmd, send_data, &recv_data, true, false,
+                           expect_empty_data, false);
+  }
+  // Msg, Cbor and Keepalive require non-empty response with status code.
+  else if (cmd == kCtapHidMsg || cmd == kCtapHidCbor ||
+           cmd == kCtapHidKeepalive) {
+    return SendRecvCommand(cmd, send_data, &recv_data, true, true, false,
+                           false);
+  }
+  // The rest are non-defined or vendor specific commands. We only send the raw
+  // message over.
+  else {
+    return SendCommand(cmd, send_data);
```

Comment:
This choice seems arbitrary to me. If we allow vendor commands, we'd need to expect an answer? If we don't call the receive function, did you check what happens to the buffered data?